### DISCUSSION TITLE
Simplest way to add content of the text file as json string field using jq🤔

### DISCUSSION THREAD

**Discussion by kloczek** [stedolan/jq#2908] [Q&A] [answered]
Lets say I have text.txt file with UTF-8 text with tabs, control characters and so on.

Q: what needs to be done to add/inject content content of that file as `{"data": "multi line text"}`?
Can I use somehow jq directly or do I need to (pre)filter that files to be able inject tat content as json field? 🤔

**Comment by jwodder** [comment]
Use the `--rawfile` option to read the file's contents into a variable, like so:

```
jq --rawfile data text.txt '{data: $data}'
```

**Reply by emanuele6** [reply]
Or just `{ $data }` instead of `{ data: $data }`

**Reply by reitzig** [reply]
NB, for others finding this through search: If you want to add _JSON_ content in a similar way, use `--slurpfile`.

**Comment by emanuele6** [comment]
If you only want to wrap the content of the file into `{"data":<content>}`:
```sh
jq -Rs '{ data: . }`
```
Otherwise, if you want to use the content of that file to modify JSON values from another input, so you can't use `-sR`, you can use `jq --rawfile var file.txt '.... $var ....' somefile.json` as suggested by @jwodder.

**Comment by kloczek** [comment]
> If you only want to wrap the content of the file into `{"data":<content>}`:
> 
> ```shell
> jq -Rs '{ data: . }`
> ```

OK it works.
Thank you.
I have yet another question: how to undo that quoting extracting that example "data" ? 🤔

**Comment by itchyny** [comment]
@kloczek Maybe you may want `jq -Rsr .` (not fully understand what _undo that quoting_ means).

**Comment by kloczek** [comment]
Simple in the "data" content for example new lines are replaced by \n. I need to extract ant present that field with replaced all those quoted parts to print original  text.

**Comment by lucas-0liveira** [comment]
To inject the content of the text.txt file as a JSON field {"data": "multi line text"}, you'll need to follow these steps:

Preprocess the Text File:

Before using jq, you may need to pre-process the text.txt file to ensure it's properly formatted for conversion to JSON.
Remove any unwanted characters, control characters, or tabs that might interfere with the JSON parsing.
Convert Text to JSON:

Once the text file is cleaned up, you can use jq to convert it into a JSON object.
If your file contains multiple lines, you might need to decide how to structure the JSON. For example, if each line should be a separate element in an array, you can use split and map.

cat text.txt | jq -R 'split("\n") | map({data: .})'


This command reads the content of text.txt, splits it into lines, and creates an array of objects with the key data for each line.

Inject the Content:

If you want to inject this JSON content into an existing JSON structure, you'll need to merge it appropriately. This can be done using tools like jq or by writing a script in a programming language like Python.
Example (assuming existing.json contains the original JSON structure):

cat existing.json | jq --argjson newContent "$(cat text.json)" '. + $newContent'

This command reads the content of existing.json, reads the content from the text file, and merges them together.

Remember, the exact commands may vary depending on the specific content and structure of your text.txt file and the existing JSON data. Always make sure to back up your data before making any modifications.

**Reply by emanuele6** [reply]
> cat text.txt | jq -R 'split("\n") | map({data: .})'

You meant `-sR`, not just `-R`.

By the way, note that rather than using slurped raw input mode, and then splitting by newline, you can more conveniently just use `-nR` and `[ inputs ]` to get an an array of the lines in the input; it also saves `jq` the time and memory of having to load the entire input in memory as a single string, to only have to split it later:
```sh
someinput | jq -nR '[ inputs | { data: . } ]'
```
or
```sh
someinput | jq -nR '[ { data: inputs } ]'
```

**Reply by kloczek** [reply]
I have yet another question.
I have difficulty with prepare params which will add content of the file in `{ "logs": { "stage1": "<Mmultiline_long_string>" } }`.
How to add such content deeper in hierarchy when when initially json string is empty? 🤔

**Comment by kloczek** [comment]
So is it possible using jq to extract such long text filed to original text form? 🤔

**Reply by wader** [reply]
You mean `jq -r .data file.json`?

**Comment by kloczek** [comment, marked as answer]
> You mean `jq -r .data file.json`?

OK it works. Thank you 👍
